backend/src/main/kotlin/cz/loono/backend/api/service/ExaminationRecordService.kt

```diff
@@ -1,139 +1,84 @@
-            throw IllegalStateException(
-                "Tried to load Examination Records for uid: $uid " +
-                    "but no such account exists."
-            )
-        }
-
-        return account.withAllRecords().examinationRecords
+    fun confirmExam(examId: Long, uid: String): ExaminationRecordDto {
+        return changeState(examId, uid, ExaminationStatusDto.CONFIRMED)
     }
 
-    private fun Account.withAllRecords(): Account {
-        val currentExamTypes = this.examinationRecords.map { record -> record.type }
-        val desiredExamTypes = ExaminationTypeEnumDto.values().map { it.name }
-
-        // Fast path = the current account already contains exactly the required examination records.
-        if (currentExamTypes.size == desiredExamTypes.size &&
-            currentExamTypes.subtract(desiredExamTypes).isEmpty()
-        ) {
-            return this
-        }
-
-        val newRecords = mutableListOf<ExaminationRecord>()
-        newRecords += retainValidRecords(examinationRecords, desiredExamTypes)
-        newRecords += createMissingRecords(this, desiredExamTypes)
-
-        deleteObsoleteRecords(examinationRecords, desiredExamTypes)
-
-        return accountRepository.save(copy(examinationRecords = newRecords))
+    @Synchronized
+    @Transactional(rollbackFor = [Exception::class])
+    fun cancelExam(examId: Long, uid: String): ExaminationRecordDto {
+        return changeState(examId, uid, ExaminationStatusDto.CANCELED)
     }
 
-    private fun retainValidRecords(
-        currentRecords: List<ExaminationRecord>,
-        desiredRecordTypes: List<String>
-    ): List<ExaminationRecord> {
-        val retainedTypes = desiredRecordTypes intersect currentRecords.map { it.type }
-
-        return currentRecords.filter { record ->
-            record.type in retainedTypes
-        }
+    fun createOrUpdateExam(examinationRecordDto: ExaminationRecordDto, uid: String): ExaminationRecordDto {
+        validateUpdateAttempt(examinationRecordDto)
+        return examinationRecordRepository.save(
+            ExaminationRecord(
+                id = examinationRecordDto.id ?: 0,
+                type = examinationRecordDto.type,
+                plannedDate = examinationRecordDto.date,
+                account = findAccount(uid),
+                firstExam = examinationRecordDto.firstExam ?: true,
+                status = examinationRecordDto.status ?: ExaminationStatusDto.NEW
+            )
+        ).toExaminationRecordDto()
     }
 
-    private fun createMissingRecords(account: Account, desiredRecordTypes: List<String>): List<ExaminationRecord> {
-        val missingTypes = desiredRecordTypes subtract account.examinationRecords.map { it.type }
-
-        return missingTypes.map { type ->
-            ExaminationRecord(type = type, account = account)
+    private fun validateUpdateAttempt(examinationRecordDto: ExaminationRecordDto) {
```

Comment:
Optional: more Kotlin idiomatic way?
```
 examinationRecordDto.id?.let {
            val record = examinationRecordRepository.findById(examinationRecordDto.id)
            if (record.isEmpty) {
                throw LoonoBackendException(
                    HttpStatus.NOT_FOUND,
                    "404",
                    "The given examination identifier not found."
                )
            }
        }
    }
```

Reply:
Spolu s prechodem na UUID prekopano. Prosim o novy check teto funkce.

Reply:
Again optional, return if-else expression? I hope you won't hate me :)
```
 private fun validateUpdateAttempt(examinationRecordDto: ExaminationRecordDto): ExaminationRecord =
        if (examinationRecordDto.uuid != null) {
            examinationRecordRepository.findByUuid(examinationRecordDto.uuid)
                ?: throw LoonoBackendException(
                    HttpStatus.NOT_FOUND,
                    "404",
                    "The given examination identifier not found."
                )
        } else {
           ExaminationRecord()
    }
 ```